presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,256 @@
+            }
+
+            pagesSpatialIndexFactory.noMoreProbeOperators();
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new SpatialJoinOperatorFactory(operatorId, planNodeId, probeTypes, probeOutputChannels, probeGeometryChannel, pagesSpatialIndexFactory);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Integer> probeOutputChannels;
+    private final int probeGeometryChannel;
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+    private final ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private final List<Type> outputTypes;
+
+    private final PageBuilder pageBuilder;
+    @Nullable
+    private Page probe;
+    private int probePosition;
+    @Nullable
+    private long[] joinAddresses;
+    private int nextJoinAddress;
+    private boolean finishing;
+    private boolean finished;
+
+    public SpatialJoinOperator(
+            OperatorContext operatorContext,
+            List<Type> outputTypes,
+            List<Type> probeTypes,
+            List<Integer> probeOutputChannels,
+            int probeGeometryChannel,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.probeTypes = ImmutableList.copyOf(probeTypes);
+        this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+        this.probeGeometryChannel = probeGeometryChannel;
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+        this.pagesSpatialIndexFuture = pagesSpatialIndexFactory.createPagesSpatialIndex();
+        this.outputTypes = ImmutableList.copyOf(outputTypes);
+        this.pageBuilder = new PageBuilder(outputTypes);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return outputTypes;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return pagesSpatialIndexFuture.isDone() && !finished && !pageBuilder.isFull() && probe == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        verify(probe == null);
+        probe = page;
+        probePosition = 0;
+
+        joinAddresses = null;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!pageBuilder.isFull() && probe != null) {
+            processProbe();
+        }
+
+        if (pageBuilder.isFull()) {
+            Page page = pageBuilder.build();
+            pageBuilder.reset();
+            return page;
+        }
+
+        if (finishing) {
+            Page page = null;
+            if (!pageBuilder.isEmpty()) {
+                page = pageBuilder.build();
+                pageBuilder.reset();
+            }
+            pagesSpatialIndexFactory.probeOperatorFinished();
```

Comment:
is this the "dereference" for the spacial index?  If so, we should clear the future since it has a hard reference to the index object.